File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {

        return help(inorder,postorder,new int[]{postorder.length-1},0,postorder.length-1);
    }
    public TreeNode help(int[] in,int[] post,int[] ind,int l,int r){
        if(l>r)return null;
        int index=find(in,post[ind[0]]);
        ind[0]-=1;
        TreeNode root=new TreeNode(in[index]);
        root.right=help(in,post,ind,index+1,r);
        root.left=help(in,post,ind,l,index-1);
        return root;
    }
    public int find(int[] arr,int val){
        for(int i=0;i<arr.length;i++){
            if(arr[i]==val)return i;
        }
        return 0;
    }
}

File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int i=0;
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        i=postorder.size()-1;
        return build(inorder,postorder,0,postorder.size()-1);
    }
    TreeNode* build(vector<int>& inorder, vector<int>& postorder, int l,int r){
        if(l>r)return NULL;
        TreeNode* newNode=new TreeNode(postorder[i]);
        int ind=search(inorder,postorder[i--]);
        newNode->right=build(inorder,postorder,ind+1,r);
        newNode->left=build(inorder,postorder,l,ind-1);
        return newNode;
    }
    int search(vector<int>& inorder,int n){
        for(int i=0;i<inorder.size();i++){
            if(inorder[i]==n)return i;
        }
        return -1;
    }

};